pkg/cache/tas_flavor_snapshot.go

```diff
@@ -296,6 +298,36 @@ func levelKey(topologyRequest *kueue.PodSetTopologyRequest) *string {
 	return nil
 }
 
+// findMostAllocatedDomainIdx binsearches a sorted array of domains and finds index of
+// the first domain with the lowest value of state, higher or equal than count.
+// If such a domain doesn't exist, it returns 0 as it's an index of the domain with the
+// most available resources
+func findMostAllocatedDomainIdx(domains []*domain, count int32) int {
+	mostAllocatedFitIdx := sort.Search(len(domains), func(i int) bool {
+		return domains[i].state < count
+	})
+	if mostAllocatedFitIdx == 0 {
+		// all domains have state < count, so we return the largest one
+		return 0
+	}
+	// find the first occurrence of the most allocated domain
+	mostAllocatedFitState := domains[mostAllocatedFitIdx-1].state
+	left, right := 0, len(domains)-1
+	for left <= right {
+		mid := left + (right-left)/2
+		switch {
+		case domains[mid].state == mostAllocatedFitState:
+			mostAllocatedFitIdx = mid
+			right = mid - 1
+		case domains[mid].state > mostAllocatedFitState:
+			left = mid + 1
+		default:
+			right = mid - 1
+		}
+	}
+	return mostAllocatedFitIdx
```

Comment:
I must say this looks like premature optimization.  This search only fires once per list domains, right? The algorithm is already over-linear O(nlogn) with the number of domains. as we already sort them per level. We will never get below linear because we need to construct the snapshot in the first place (unless I'm missing something).

If we really need binary search let's use library rather than re-implement it :)